File: sample/btree/pathK/pathK.go
// 2分木における経路の和がK
package main

import "fmt"
import . "../../btree"

func pathK(n *Node, K int) {
	path := []int{}
	var dfs func(*Node)
	dfs = func(n *Node) {
		if n == nil {
			return
		}
		path = append(path, n.Value)
		for i := 0; i < len(path); i++ {
			s := 0
			for j := i; j < len(path); j++ {
				s += path[j]
			}
			if s == K {
				fmt.Println(path[i:])
			}
		}
		dfs(n.Left)
		dfs(n.Right)
		path = path[:len(path)-1] // 親に戻る時は子の経路を排除
	}
	dfs(n)
}

func main() {
    bt := FromSlice([]int{0, 1, 2, 3, 4, 5, 6})
    fmt.Println(bt)
	pathK(bt, 6)
}

File: sample/btree/pathK/pathK.hs
-- 2分木における経路の和がK

data Tree a = Leaf | Node a (Tree a) (Tree a)
  deriving Show

pathK :: Num a => Tree a -> a -> [[a]]
pathK n k = go n [] where
  go Leaf _ = []
  go n@(Node p l r) path = a ++ b ++ c where
    a = go l (p:path)
    b = go r (p:path)
    c = (p:path)

fromList :: [Int] -> Tree Int
fromList [] = Leaf
fromList (x:xs) = Node x (fromList ys) (fromList zs) where
  (ys, zs) = split xs

split (x:y:zs) = (x:xs, y:ys) where (xs, ys) = split(zs)
split x = (x, [])

main = do
  print $ pathK (fromList [0..6]) 6
